source/extensions/clusters/aggregate/cluster.cc

```diff
@@ -0,0 +1,148 @@
+            host_set->localityWeights(), host_set->hosts(), {}, host_set->overprovisioningFactor());
+        priority_to_cluster.emplace_back(std::make_pair(priority_in_current_cluster, tlc));
+      }
+      priority_in_current_cluster++;
+    }
+  }
+
+  return std::make_pair(std::move(priority_set), std::move(priority_to_cluster));
+}
+
+void Cluster::startPreInit() {
+  for (const auto& cluster : clusters_) {
+    auto tlc = cluster_manager_.get(cluster);
+    if (tlc == nullptr) {
+      continue;
+    }
+
+    // Add callback for clusters initialized before aggregate cluster.
+    tlc->prioritySet().addMemberUpdateCb(
+        [this, cluster](const Upstream::HostVector&, const Upstream::HostVector&) {
+          ENVOY_LOG(info, "member update for cluster '{}'", cluster);
+          refresh();
+        });
+  }
+  refresh();
+  handle_ = cluster_manager_.addThreadLocalClusterUpdateCallbacks(*this);
+
+  onPreInitComplete();
+}
+
+void Cluster::refresh(std::function<bool(const std::string&)>&& deleting) {
+  auto pair = linearizePrioritySet(std::move(deleting));
+  if (pair.first.hostSetsPerPriority().empty()) {
+    if (load_balancer_ != nullptr) {
+      load_balancer_.reset();
+    }
+  } else {
+    load_balancer_ = std::make_shared<LoadBalancerImpl>(*this, std::move(pair));
+  }
+}
+
+void Cluster::onClusterAddOrUpdate(Upstream::ThreadLocalCluster& cluster) {
+  if (std::find(clusters_.begin(), clusters_.end(), cluster.info()->name()) != clusters_.end()) {
+    ENVOY_LOG(info, "update or add cluster '{}'", cluster.info()->name());
+    refresh();
+    cluster.prioritySet().addMemberUpdateCb(
+        [this](const Upstream::HostVector&, const Upstream::HostVector&) { refresh(); });
+  }
+}
+
+void Cluster::onClusterRemoval(const std::string& cluster_name) {
+  //  The onClusterRemoval callback is called before the thread local cluster was removed. There
+  //  will be a dangling pointer to the thread local cluster if delete cluster is not skipped.
+  if (std::find(clusters_.begin(), clusters_.end(), cluster_name) != clusters_.end()) {
+    ENVOY_LOG(info, "remove cluster '{}'", cluster_name);
+    refresh([&cluster_name](const std::string& c) { return cluster_name == c; });
+  }
+}
+
+Upstream::HostConstSharedPtr
+Cluster::LoadBalancerImpl::chooseHost(Upstream::LoadBalancerContext* context) {
+  const auto priority_pair =
+      choosePriority(random_.random(), per_priority_load_.healthy_priority_load_,
+                     per_priority_load_.degraded_priority_load_);
+  AggregateLoadBalancerContext aggregate_context(context, priority_pair.second,
+                                                 priority_to_cluster_[priority_pair.first].first);
+  return priority_to_cluster_[priority_pair.first].second->loadBalancer().chooseHost(
+      &aggregate_context);
+}
+
+Upstream::HostConstSharedPtr
+AggregateClusterLoadBalancer::chooseHost(Upstream::LoadBalancerContext* context) {
+  auto load_balancer = cluster_.getLoadBalancer();
```

Comment:
i think this ends up being unsafe since you're accessing the cluster that is updated on the main thread from the worker threads. I'd suggest taking a look at how the ClusterManager uses a ClusterEntry and posts changes to a cluster to to all the worker threads, having the priority set that is actually used for load balancing live on a ThreadLocalCluster instead of the main Cluster. The flow for those updates are basically: 1) change happens on the main thread 2) PrioritySet on the main thread is updated 3) update callbacks are triggered that post the new state to the worker threads using `postThreadLocalClusterUpdate`. Alternatively you could follow the docs on `ThreadAwareLoadBalancer` to make this work without posting to the workers, but I suspect handling that will end up being more complicated than making this thread local.

given this complexity we probably want to add an integration test that spins up an envoy instance and sends some requests through the aggregate cluster

Reply:
As discussed offline. Implemented the update using TLS post.